src/clients/HttpTunneler.cc

```diff
@@ -122,23 +123,66 @@ Http::Tunneler::startReadingResponse()
 }
 
 void
-Http::Tunneler::writeRequest()
+Http::Tunneler::buildHttpHeaders(HttpHeader &hdr_out)
 {
-    debugs(83, 5, connection);
-
-    HttpHeader hdr_out(hoRequest);
     Http::StateFlags flags;
     flags.peering = true;
     // flags.tunneling = false; // the CONNECT request itself is not tunneled
     // flags.toOrigin = false; // the next HTTP hop is a non-originserver peer
+    if (request->method == Http::METHOD_CONNECT) {
+        // We are forwarding an HTTP CONNECT request
+        HttpStateData::httpBuildRequestHeader(request.getRaw(),
+                                              nullptr, // StoreEntry
+                                              al,
+                                              &hdr_out,
+                                              flags);
+    } else {
+        // We need to create a new CONNECT request to forward the HttpRequest object
+
+        // The new HttpRequest object shares the same features with the
+        // origin, including the masterXaction object.
+        // XXX: Should the peerRequest be handled as a new HTTP transaction?
+        HttpRequest::Pointer peerRequest = new HttpRequest(request->masterXaction);
+        peerRequest->initHTTP(Http::METHOD_CONNECT, AnyP::PROTO_HTTP, "http", url.c_str());
+
+        // OK about auth and client related info, but what about
+        // adaptHistory_/icapHistory_ ?
+        Http::Message *msg = peerRequest.getRaw();
+        msg->inheritProperties(request.getRaw());
+
+        AccessLogEntry::Pointer peerRequestAl = new AccessLogEntry;
```

Comment:
This CONNECT request is a Squid-server HTTP transaction for an existing master transaction. We are not going to log this CONNECT request. ALE is only created for new master transactions (which are logged to access.log) and, IIRC, for adaptation transactions (because they have their own equivalent of access.log). So why are we creating a new ALE here?

Reply:
I decided to build a new ALE because of FilledCheckList::syncAle. Here the al->request is updated to point to the passed HttpRequest object. This is will be updated to point to the CONNECT HttpRequest object and the AccessLogEntry will point to wrong HttpRequest  object until a new syncAle call update it to point to the correct HttpRequest.

Reply:
> FilledCheckList::syncAle [updates al->request] to point to the passed HttpRequest object. This is will be updated to point to the CONNECT HttpRequest object and the AccessLogEntry will point to wrong HttpRequest object

OK, so we are selecting between the following two evils here:

1. the correct ALE starts pointing to the wrong (outgoing CONNECT) request object;
2. ACL checks inside httpBuildRequestHeader() start using the wrong ALE.

You picked 2 and tried to mitigate the "wrong ALE" problem by copying the correct ALE fields to the new ALE. However, the ALE class is not meant to be cloned or copied. Even if we fix the new ALE-copying code to call some new ALE::inheritProperties() method (to ensure that nothing is forgotten during copying), some httpBuildRequestHeader() code or even individual ACLs may update the new/wrong ALE; those updates will be lost. Adaptation aside, there should be only one ALE per master transaction, and ALE-using code may rely on that invariant.

It is not clear to me whether the syncAle() call is conceptually wrong (i.e. it should not assume that the request is the client request) or this PR assertion that httpBuildRequestHeader() ALCs should look at the _outgoing_ request is wrong, but I suspect it is the latter.

The root of the problem here could be in the fact that while most ACL-driven directives work with _incoming_ (including adapted) requests, the request_header_access directive works with _outgoing_ requests. That is a rare (if not the only) exception AFAICT.

I wonder if the code we are trying to fix is actually not broken: Yes, request_header_access is used to filter outgoing request headers. However, that does not mean that its ACLs should have access to the outgoing request! The directive asks an "Am I allowed to send request header field named Foo?" question. In most cases, the answer can and _should_ be given _without_ looking at the outgoing request being formed!

In other words, instead of trying to "fix" code so that request_header_access ACLs can match the outgoing CONNECT, we should claim that the code is working correctly. There is just no ACL that answers an "Am I building a CONNECT request?" question. We can either add such an ACL or suggest how to detect that condition using existing ACLs. Based on your work for this issue, do you think such detection is possible without adding a dedicated ACL?

If such detection is not possible, we could suggest adding a building_connect_request ACL that would only match inside Http::Tunneler::writeRequest(). A more general solution would be to extend the existing at_step ACL to add support for a new "building_connect_request" step (not directly related to SslBump!). IIRC, the at_step ACL was specifically named so that it can be extended beyond SslBump steps. It does not have an "ssl::" scheme and, based on the current documentation, can be safely extended beyond SslBump to add other processing stages. The stages may even overlap.

What do you think about adding a building_connect_request ACL  or extending at_step as the correct solution for this issue?

> until a new syncAle call update it to point to the correct HttpRequest.

_If_ you find flaws in my logic in the beginning of this comment, then we will need to discuss whether calling syncAle() twice (once automatically in the old code and once "manually" in Tunneler::buildHttpHeaders() is better than copying ALE.

Reply:
> OK, so we are selecting between the following two evils here:
> 
>     1. the correct ALE starts pointing to the wrong (outgoing CONNECT) request object;
> 
>     2. ACL checks inside httpBuildRequestHeader() start using the wrong ALE.
> 
> 
> You picked 2 and tried to mitigate the "wrong ALE" problem by copying the correct ALE fields to the new ALE. However, the ALE class is not meant to be cloned or copied. 

Well, yep.

> 
> It is not clear to me whether the syncAle() call is conceptually wrong (i.e. it should not assume that the request is the client request) or this PR assertion that httpBuildRequestHeader() ALCs should look at the _outgoing_ request is wrong, but I suspect it is the latter.

I agree here.

> 
> The root of the problem here could be in the fact that while most ACL-driven directives work with _incoming_ (including adapted) requests, the request_header_access directive works with _outgoing_ requests. That is a rare (if not the only) exception AFAICT.

This is not true in current master branch. The request_header_access currently works with incoming request.  This patch try to make it work with outgoing requests.

> 
> In other words, instead of trying to "fix" code so that request_header_access ACLs can match the outgoing CONNECT, we should claim that the code is working correctly. There is just no ACL that answers an "Am I building a CONNECT request?" question.

I agree.

>  We can either add such an ACL or suggest how to detect that condition using existing ACLs. Based on your work for this issue, do you think such detection is possible without adding a dedicated ACL?

Unfortunately looks impossible.

> If such detection is not possible, we could suggest adding a building_connect_request ACL that would only match inside Http::Tunneler::writeRequest(). 

Does not looks bad.

> A more general solution would be to extend the existing at_step ACL to add support for a new "building_connect_request" step (not directly related to SslBump!). IIRC, the at_step ACL was specifically named so that it can be extended beyond SslBump steps. It does not have an "ssl::" scheme and, based on the current documentation, can be safely extended beyond SslBump to add other processing stages. The stages may even overlap.

Probably better then the building_connect_request ACL. In any case they are looking similar.

> 
> What do you think about adding a building_connect_request ACL or extending at_step as the correct solution for this issue?

Yes it solves our problem.
However my sense is that probably we have more issues to solve, not just identifying the outgoing CONNECT request. Imagine a CONNET request is received by squid which includes the custom header 'X-Custom'. Currently in squid:
1) If this request is not bumped and forwarded to a peer the X-Custom header will be included in the peer CONNECT  request.
2) If the request bumped after step2 or spliced the X-Custom header will be included in the peer CONNECT request
3) If the request bumped after step1(client-first bumping) the X-Custom header will NOT be included in the peer CONNECT request, and there is not any way to add it.

Probably there are other similar problems, where features of the original CONNECT request are dropped (for example authentication info, I need to check more on this). So what about the following: Store the ALE and HttpRequest object of the client CONNECT request to squid in the case of client-first bumping and use it to generate outgoing request.

In any case a building_connect_request ACL still needed because in future would have to handle more "GET https://" cases (not only those coming from client-first bumping mode).

Reply:
> Alex: The root of the problem here could be in the fact that while most ACL-driven directives work with incoming (including adapted) requests, the request_header_access directive works with outgoing requests. That is a rare (if not the only) exception AFAICT.

> Christos: This is not true in current master branch. The request_header_access currently works with incoming request. This patch try to make it work with outgoing requests.

To clarify: I was not talking about how the code is or was written. I was talking about how the admin perceives squid.conf directives. The admin either assumes that the directive ACLs are going to examine the incoming (including adapted) request or that they are going to examine the outgoing request. In most (possibly all other) cases, the former assumption is both natural and correct. However, In case of request_header_access, an admin may _naturally_ think about the _outgoing_ request (e.g., the customer did). That natural assumption may be _wrong_ -- the code was using and should _continue_ to use the incoming request instead.

I am digesting your other responses and plan to suggest a specific course of action soon.

Reply:
> If this CONNECT request is not bumped and forwarded to a peer the X-Custom header will be included in the peer CONNECT request.

That is a bug, right? CONNECT is a hop-by-hop mechanism. Squid should not forward hop-by-hop info (unless explicitly configured to do so for, say, authentication purposes).

> If the CONNECT request bumped after step2 or spliced, the X-Custom header will be included in the peer CONNECT request

Also a bug, for the same reason, right?


> If the request client-first bumped, the X-Custom header will NOT be included in the peer CONNECT request, 

Sounds like the right behavior to me, for the same hop-by-hop reasons.

> and there is not any way to add it

How about this:

```
acl generatingCONNECT at_step GeneratingCONNECT
acl haveCustom note ReceivedCustom
request_header_add X-Custom "%note{ReceivedCustom}" generatingCONNECT haveCustom
```

... where the ReceivedCustom connection annotation was set to the received X-Custom value when the received CONNECT request was processed/allowed.

Reply:
> That is a bug, right? CONNECT is a hop-by-hop mechanism. Squid should not forward hop-by-hop info (unless explicitly configured to do so for, say, authentication purposes).

Well, not sure about this. The mozilla developer manual says that it is a hop-by-hop method:
  https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/CONNECT
However the rfc7231 says:
```
   The recipient proxy can establish a tunnel either by directly
   connecting to the request-target or, if configured to use another
   proxy, by forwarding the CONNECT request to the next inbound proxy.
```
The proxy has to forward the CONNECT request of cource removing hop-by-hop headers.
In our case, we terminated the CONNECT request, we are creating new CONNECT request is definitely a hop-by-hop. But it has to do with what the users expect.

Reply:
Interesting analysis, thank you! I [asked](https://lists.w3.org/Archives/Public/ietf-http-wg/2019AprJun/0051.html) HTTP WG for a clarification. Even if the consensus is end-to-end, we may not implement that behavior immediately, but it would be good to know what we should be aiming for long-term.